File: cpp/2353.design-a-food-rating-system.cpp
class FoodRatings {
public:
    unordered_map<string, set<pair<int, string> > >cuisine2scores;
    unordered_map<string, string> food2cuisine;
    unordered_map<string, int> food2score;
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
	    for(int i = 0 ; i < foods.size(); i++){
		    food2score[foods[i]] = ratings[i];
		    food2cuisine[foods[i]] = cuisines[i];
		    cuisine2scores[cuisines[i]].insert({-ratings[i], foods[i]});
	    }
    }
    
    void changeRating(string food, int newRating) {
	    auto cuisine = food2cuisine[food];
	    cuisine2scores[cuisine].erase({-food2score[food], food});
	    food2score[food] = newRating;
	    cuisine2scores[cuisine].insert({-food2score[food], food});
    }
    
    string highestRated(string cuisine) {
	    return cuisine2scores[cuisine].begin()->second;
    }
};

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings* obj = new FoodRatings(foods, cuisines, ratings);
 * obj->changeRating(food,newRating);
 * string param_2 = obj->highestRated(cuisine);
 */

File: python/2353.design-a-food-rating-system.py
class FoodRatings:


    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.cuisine2foods = {}
        self.food2rating = {}
        self.food2cuisine = {}

        for i in range(len(foods)):
            self.food2rating[foods[i]] = ratings[i]
            self.food2cuisine[foods[i]] = cuisines[i]
            if cuisines[i] not in self.cuisine2foods:
                self.cuisine2foods[cuisines[i]] = sortedcontainers.SortedSet()
            self.cuisine2foods[cuisines[i]].add((-ratings[i], foods[i]))


    def changeRating(self, food: str, newRating: int) -> None:
        cuisine = self.food2cuisine[food]
        self.cuisine2foods[cuisine].remove((-self.food2rating[food], food))
        self.food2rating[food] = newRating
        self.cuisine2foods[cuisine].add((-self.food2rating[food], food))


    def highestRated(self, cuisine: str) -> str:
        return self.cuisine2foods[cuisine][0][1]


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)
